fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,103 @@
+import TodoForm from './todo-form';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     */
+    fetch('/api/todos', { method: 'GET' })
+      .then(res => res.json())
+      .then(data => this.setState({ todos: data }));
+  }
+
+  addTodo(newTodo) {
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    * add the created todo to the state array.
+    *
+    * Do not mutate the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+    fetch('/api/todos',
+      {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify(newTodo)
+      })
+      .then(res => res.json())
+      .then(data => {
+        const newData = this.state.todos.concat(data);
+        this.setState({ todos: newData });
+      });
+  }
+
+  toggleCompleted(todoId) {
+    /**
+     * Find the index of the todo with the matching todoId in the state array.
+     * Get its "isCompleted" status.
+     * Make a new object containing the opposite "isCompleted" status.
+     * Use fetch to send a PATCH request to `/api/todos/${todoId}`
+     * Then 😉, once the response JSON is received and parsed,
+     * replace the old todo in the state array.
+     *
+     * NOTE: "toggle" means to flip back and forth, so clicking a todo
+     * in the list should "toggle" its isCompleted status back and forth.
+     *
+     * Do not mutate the original state array, nor any objects within it.
+     * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+     *
+     * TIP: Be sure to SERIALIZE the updates in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+    const current = this.state.todos[todoId - 1];
+    if (current.isCompleted) current.isCompleted = false;
+    else current.isCompleted = true;
+    fetch(`/api/todos/${current.todoId}`,
+      {
+        method: 'PATCH',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify(current)
+      })
+      .then(res => res.json())
+      .then(data => {
+        const all = this.state.todos;
```

Comment:
Arrays are reference data types so this won't copy the array it will just create a new pointer to the same array. You can use `slice` or the ES6 spread operator to create a shallow copy of the array.

```JAVASCRIPT
// Using .slice
const newArray = this.state.todos.slice();

// Using ES6 spread operator
const newArray = [ ...this.state.todos ];
```